### PR TITLE
generate OpenAPI schema for Helm schematic compDef

### LINKED ISSUE #1222: generate OpenAPI schema for helm chart values
**Is your feature request related to a problem? Please describe.**

Currently, we have [Helm schematic workloads](https://github.com/oam-dev/kubevela/blob/master/docs/en/helm/basic.md), which allows to vela install a Helm release into the cluster. 

An application containing a Helm-based workload will use `.spec.settings` to override default Chart values. That's similar to how [parameter](https://kubevela.io/#/en/platform-engineers/workload-type?id=_4-define-template) works for a CUE-based workload. Therefore, Helm schematic workloads should provide an OpenAPI v3 JSON Schema in the definition for the same purpose introduced [in this doc](https://github.com/oam-dev/kubevela/blob/master/docs/en/platform-engineers/openapi-v3-json-schema.md).

Helm has support for json schema validation, for those Helm charts that already provide a  `values.schema.json` file, we can just re-use it. But for most majority having no schema file, Vela should generate one according to the `values.yml`.

**Describe the solution you'd like**

In ComponentDefinitionController
- Resolve the Helm schematic in workloadDef to get the info of Chart
- Fetch Chart bundle from remote
- Generate JSON schema based on values.yaml through [this tool](https://github.com/karuppiah7890/helm-schema-gen)

Attention: [This tool](https://github.com/karuppiah7890/helm-schema-gen) only generates a JSON schema scratch but not a true JSON schema. Community has no perfect solution for generating JSON schema from values.yaml automatically.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by captainroy-hy**
to fix #1222 

Signed-off-by: roy wang <seiwy2010@gmail.com>

### CODE REVIEW

File: pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller.go [hunk 1]
```diff
@@ -86,6 +86,10 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	if workloadType == util.ReferWorkload {
 		def.WorkloadDefName = componentDefinition.Spec.Workload.Type
 	}
+	if workloadType == util.HELMDef {
+		def.Helm = componentDefinition.Spec.Schematic.HELM
+		def.ComponentDefinition = componentDefinition
+	}
```
Comment (captainroy-hy) [inline pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller.go:92]: Currently just make minimal changes in CompDef controller in this PR. 
I would like to refactor it latter.

Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/core/components/componentdefinition/componentdefinition_controller.go:92]: please write a `switch/case` here

File: pkg/controller/utils/capability.go [hunk 2]
```diff
@@ -105,7 +106,13 @@ func (def *CapabilityComponentDefinition) GetOpenAPISchema(ctx context.Context,
...
 // StoreOpenAPISchema stores OpenAPI v3 schema in ConfigMap from WorkloadDefinition
 func (def *CapabilityComponentDefinition) StoreOpenAPISchema(ctx context.Context, k8sClient client.Client, namespace, name string) error {
-	jsonSchema, err := def.GetOpenAPISchema(ctx, k8sClient, namespace, name)
+	var jsonSchema []byte
+	var err error
+	if def.WorkloadType == util.HELMDef {
+		jsonSchema, err = helm.GetChartValuesJSONSchema(ctx, def.Helm)
+	} else {
+		jsonSchema, err = def.GetOpenAPISchema(ctx, k8sClient, namespace, name)
+	}
```
Comment (wonderflow) [inline pkg/controller/utils/capability.go:115]: a switch/default is better

File: pkg/appfile/helm/schema_test.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestGenerateSchemaFromValues(t *testing.T) {
```
Comment (wonderflow) [inline pkg/appfile/helm/schema_test.go:12]: good job, it'll be better if we can have more cases with small different schema cases file.

File: pkg/appfile/helm/schema.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+	files, err := loader.LoadArchiveFiles(resp.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot load Chart files")
+	}
+	return files, nil
+}
+
+// cue openapi encoder converts default in Chart Values as enum in schema
+// changing enum to default makes the schema consistent with Chart Values
+func changeEnumToDefault(schema *openapi3.Schema) {
```
Comment (wonderflow) [inline pkg/appfile/helm/schema.go:150]: let's add a test for this function

File: pkg/appfile/helm/schema.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+	b, err := json.Marshal(m)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot marshall schema")
+	}
+	return b, nil
+}
+
+// handleItemsOfArrayType will convert all 'items' of array type from array to object
+// and remove enum in the items
+func handleItemsOfArrayType(t map[string]interface{}) {
```
Comment (wonderflow) [inline pkg/appfile/helm/schema.go:190]: add a test for this function

File: pkg/appfile/helm/schema_test.go (hunk 3 above)
Comment (captainroy-hy) [reply pkg/appfile/helm/schema_test.go:12]: Yes, that's what we really need. The current `yaml=>cue=>openapi` approach may fail for special cases we haven't tested.

File: pkg/appfile/helm/schema.go (hunk 4 above)
Comment (captainroy-hy) [reply pkg/appfile/helm/schema.go:150]: got it.

File: pkg/appfile/helm/schema.go (hunk 5 above)
Comment (captainroy-hy) [reply pkg/appfile/helm/schema.go:190]: got it.

File: docs/en/helm/component.md [hunk 6]
```diff
@@ -68,8 +68,17 @@ spec:
...
 ```
 Helm module workload will use data in `settings` as [Helm chart values](https://github.com/captainroy-hy/podinfo/blob/master/charts/podinfo/values.yaml).
-Currently, you can learn the schema of settings by reading the `README.md` of the Helm chart, and the schema are totally align with [`values.yaml`](https://github.com/captainroy-hy/podinfo/blob/master/charts/podinfo/values.yaml) of the chart.
-We will integerate the values and generate it's [openapi-v3-json-schema](https://kubevela.io/#/en/platform-engineers/openapi-v3-json-schema.md) soon.
+You can learn the schema of settings by reading the `README.md` of the Helm
+chart, and the schema are totally align with
+[`values.yaml`](https://github.com/captainroy-hy/podinfo/blob/master/charts/podinfo/values.yaml)
+of the chart.  
+
+Additionally, Vela will fetch the `values.schema.json` file from the Chart
```
Comment (wonderflow) [inline docs/en/helm/component.md:76]: In which case will the `values.schema.json` exist in a helm chart? Is there any standard ?

File: pkg/appfile/helm/helm.go [hunk 7]
```diff
@@ -5,13 +5,13 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/pkg/errors"
+
```
Comment (wonderflow) [inline pkg/appfile/helm/helm.go:9]: remove the empty line

File: pkg/appfile/helm/helm.go [hunk 8]
```diff
@@ -102,3 +98,15 @@ func setSpecObjIntoUnstructuredObj(spec interface{}, u *unstructured.Unstructure
 	_ = unstructured.SetNestedMap(u.Object, data, "spec")
 	return nil
 }
+
+func unmarshalHelmSpec(h *v1alpha2.Helm) (*helmapi.HelmReleaseSpec, *helmapi.HelmRepositorySpec, error) {
```
Comment (wonderflow) [inline pkg/appfile/helm/helm.go:102]:
```suggestion
func decodeHelmSpec(h *v1alpha2.Helm) (*helmapi.HelmReleaseSpec, *helmapi.HelmRepositorySpec, error) {
```

File: pkg/appfile/helm/schema.go [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+	var values, jsonSchema *loader.BufferedFile
+	for _, f := range files {
+		if f.Name == "values.yaml" || f.Name == "values.yml" {
+			values = f
+			continue
+		}
+		if f.Name == "values.schema.json" {
+			jsonSchema = f
+			continue
+		}
```
Comment (wonderflow) [inline pkg/appfile/helm/schema.go:54]: from the code style aspect, it's better to use a `switch`. And merge the logic L56~59 into it.

File: pkg/appfile/helm/schema.go [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+	}
+	if jsonSchema != nil {
+		// use the JSON schema file if exists
+		return jsonSchema.Data, nil
+	}
+	if values == nil {
+		return nil, errors.New("cannot find JSON schema nor Values file in the Chart")
```
Comment (wonderflow) [inline pkg/appfile/helm/schema.go:61]:
```suggestion
		return nil, errors.New("cannot find 'values.schema.json' or 'values.yaml' file in the Chart")
```

File: docs/en/helm/component.md (hunk 6 above)
Comment (captainroy-hy) [reply docs/en/helm/component.md:76]: I will add [Helm official doc](https://helm.sh/docs/topics/charts/#schema-files) here.

**Review by wonderflow — APPROVED** [review]
Good Job

**Pull request merged**